### PR TITLE
[ConstraintSystem] Relax the left-over information check in `ComponentStep`

### PR DESCRIPTION
**Pull request by xedin**
Instead of crashing in release builds, let's simply fail the
step and let type-checker produce a fallback diagnostic.

Resolves: rdar://56167233

### CODE REVIEW

File: lib/Sema/CSStep.cpp [hunk 1]
```diff
@@ -362,22 +362,60 @@ StepResult ComponentStep::take(bool prevFailed) {
...
       llvm::report_fatal_error("Free type variables left over?");
+#else
+      if (CS.isDebugMode()) {
+        getDebugLogger() << "(failed due to remaining free type variables)\n";
+        CS.print(getDebugLogger(/*indent=*/false));
+      }
+
+      return finalize(/*isSuccess=*/false);
+#endif
     }
```
Comment (hamishknight) [inline lib/Sema/CSStep.cpp:419]: I think this entire condition is actually dead code, as it's already handled in an above `if` statement, I guess I must have missed that when adding it

**Review by hamishknight — APPROVED** [review]

File: lib/Sema/CSStep.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSStep.cpp:419]: Ah, I see yeah, let's remove it.

**Comment by xedin** [comment]
@hamishknight I have removed redundant check you mentioned and added `InvalidState` flag to make sure that solving is aborted early and solutions are never produced, this should make sure that solving in release mode always ends up with a fallback diagnostic.

File: lib/Sema/CSStep.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSStep.cpp]: Fixed!

**Review by hamishknight — APPROVED** [review]
Thanks!

**Review by hborla — APPROVED** [review]
Good idea to produce an error message instead 👍

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@hamishknight @hborla After thinking about this some more I think we can drop abort blocks all together because there it would be impossible now to miscompile, WDYT?

**Comment by hborla** [comment]
@xedin Yeah, I think it's fine to always report the error message instead of aborting. If we still want it to crash in this part of the code in a debug build, we could just put in an assert

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
Adjusted that, for debugging it would print the constraints in question so I just removed NDEBUG blocks.

**Comment by hamishknight** [comment]
I'd prefer if we crashed in a debug build to make it loud and obvious that we have a logic error, but yeah that could be done with an assert.

**Comment by xedin** [comment]
I think asking to file a bug is loud enough, that would also indicate where the error is better than crash report because not all tools capture stacktraces.

**Comment by hamishknight** [comment]
Hmm, won't we emit "type of expression is ambiguous without more context" for an expression with no solutions?

**Comment by xedin** [comment]
Ugh, you are right! I have completely forgot about that message assuming that `forUndiagnosedError` would always result in `failed to produce a diagnostic ...` message. Fixed now.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test

**Pull request merged**